File: java/integrations/okhttp/src/main/java/io/zrz/httpsig/okhttp/SigningInterceptor.java
package io.zrz.httpsig.okhttp;

import java.io.IOException;
import java.util.function.Function;

import io.zrz.httpsig.HttpSigException;
import io.zrz.httpsig.SignatureParameters;
import io.zrz.httpsig.Signer;
import io.zrz.httpsig.SigningKey;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * OkHttp {@link Interceptor} that signs outgoing requests with HTTP Message Signatures (RFC 9421).
 *
 * <pre>
 *   var interceptor = new SigningInterceptor(
 *       signingKey,
 *       req -> SignatureParameters.builder()
 *           .component("@method")
 *           .component("@authority")
 *           .keyId("my-key")
 *           .created(Instant.now())
 *           .build()
 *   );
 *
 *   var client = new OkHttpClient.Builder()
 *       .addInterceptor(interceptor)
 *       .build();
 * </pre>
 */
public final class SigningInterceptor implements Interceptor {

    private final SigningKey key;
    private final String label;
    private final Function<Request, SignatureParameters> paramsFactory;

    /**
     * @param key           the signing key
     * @param paramsFactory produces {@link SignatureParameters} for each request
     */
    public SigningInterceptor(SigningKey key, Function<Request, SignatureParameters> paramsFactory) {
        this(key, "sig1", paramsFactory);
    }

    /**
     * @param key           the signing key
     * @param label         the signature label (e.g. "sig1")
     * @param paramsFactory produces {@link SignatureParameters} for each request
     */
    public SigningInterceptor(SigningKey key, String label, Function<Request, SignatureParameters> paramsFactory) {
        this.key = key;
        this.label = label;
        this.paramsFactory = paramsFactory;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request original = chain.request();

        SignatureParameters params = paramsFactory.apply(original);
        var msg = new OkHttpMessage(original);

        Signer.SignResult result;
        try {
            result = Signer.sign(msg, label, params, key, null);
        } catch (HttpSigException e) {
            throw new IOException("failed to sign request", e);
        }

        Request signed = original.newBuilder()
            .addHeader("Signature-Input", Signer.signatureInputHeader(result))
            .addHeader("Signature", Signer.signatureHeader(result))
            .build();

        return chain.proceed(signed);
    }
}

File: kotlin/integrations/okhttp/src/main/kotlin/com/zourzouvillys/httpsig/okhttp/SigningInterceptor.kt
package com.zourzouvillys.httpsig.okhttp

import com.zourzouvillys.httpsig.SignatureParameters
import com.zourzouvillys.httpsig.Signer
import com.zourzouvillys.httpsig.SigningKey
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response

/**
 * OkHttp [Interceptor] that signs outgoing requests per RFC 9421.
 *
 * Adds `Signature-Input` and `Signature` headers to every request that passes through.
 *
 * @param key the signing key to use
 * @param label the signature label (defaults to "sig1")
 * @param paramsFactory builds [SignatureParameters] for each request, so callers can
 *   vary components or metadata per-request if needed
 */
class SigningInterceptor(
    private val key: SigningKey,
    private val label: String = "sig1",
    private val paramsFactory: (Request) -> SignatureParameters,
) : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val original = chain.request()
        val msg = OkHttpMessage(original)
        val params = paramsFactory(original)
        val result = Signer.sign(msg, label, params, key)

        val signed = original.newBuilder()
            .header("Signature-Input", Signer.signatureInputHeader(result))
            .header("Signature", Signer.signatureHeader(result))
            .build()

        return chain.proceed(signed)
    }
}
